File: yarl-java/src/test/java/hr/tjakopan/yarl/ContextTest.java
package hr.tjakopan.yarl;

import org.junit.Test;

import java.util.HashMap;

import static org.assertj.core.api.Assertions.assertThat;

public class ContextTest {
  @Test
  public void shouldAssignOperationKeyFromConstructor() {
    final var context = Context.of("SomeKey");

    assertThat(context.getOperationKey()).isEqualTo("SomeKey");
    assertThat(context.keySet().size()).isEqualTo(0);
  }

  @Test
  public void shouldAssignOperationKeyAndContextDataFromConstructor() {
    final var context = Context.of("SomeKey", new HashMap<>() {{
      put("key1", "value1");
      put("key2", "value2");
    }});

    assertThat(context.getOperationKey()).isEqualTo("SomeKey");
    assertThat(context.get("key1")).isEqualTo("value1");
    assertThat(context.get("key2")).isEqualTo("value2");
  }

  @Test
  public void noArgsConstructorShouldAssignNoOperationKey() {
    final var context = Context.of();

    assertThat(context.getOperationKey()).isNull();
  }

  @Test
  public void shouldAssignCorrelationIdWhenAccessed() {
    final var context = Context.of("SomeKey");

    assertThat(context.getCorrelationId()).isNotNull();
  }

  @Test
  public void shouldReturnConsistentCorrelationId() {
    final var context = Context.of("SomeKey");

    final var uuid1 = context.getCorrelationId();
    final var uuid2 = context.getCorrelationId();

    assertThat(uuid1).isSameAs(uuid2);
  }
}

File: yarl-core/src/test/kotlin/hr/tjakopan/yarl/ContextTest.kt
package hr.tjakopan.yarl

import org.assertj.core.api.Assertions.assertThat
import kotlin.test.Test

class ContextTest {
  @Test
  fun `should assign operation key from constructor`() {
    val context = Context(operationKey = "SomeKey")

    assertThat(context.operationKey).isEqualTo("SomeKey")
    assertThat(context.keys.size).isEqualTo(0)
  }

  @Test
  fun `should assign operation key and context data from constructor`() {
    val context = Context(operationKey = "SomeKey", contextData = mutableMapOf("key1" to "value1", "key2" to "value2"))

    assertThat(context.operationKey).isEqualTo("SomeKey")
    assertThat(context["key1"]).isEqualTo("value1")
    assertThat(context["key2"]).isEqualTo("value2")
  }

  @Test
  fun `no-args constructor should assign no operation key`() {
    val context = Context()

    assertThat(context.operationKey).isNull()
  }

  @Test
  fun `should assign correlation id when accessed`() {
    val context = Context(operationKey = "SomeKey")

    @Suppress("UsePropertyAccessSyntax")
    assertThat(context.correlationId).isNotNull()
  }

  @Test
  fun `should return consistent correlation id`() {
    val context = Context(operationKey = "SomeKey")

    val uuid1 = context.correlationId
    val uuid2 = context.correlationId

    assertThat(uuid1).isSameAs(uuid2)
  }
}
